File: ReactlikeMvvmVb/HiCustomCommand/HiCommand.vb
Imports System
Imports System.Threading.Tasks
Imports System.Windows.Input

Public Class HiCommand
    Implements ICommand
    Private Property _fnCommand As Action
    Private Property _fnCommandAsync As Func(Of Task)

    Public ReadOnly Property EAsync As Boolean
        Get
            Return _fnCommand Is Nothing
        End Get
    End Property

    Public Sub New(fnCommand As Action)
        _fnCommand = fnCommand
    End Sub

    Public Sub New(fnCommandAsync As Func(Of Task))
        _fnCommandAsync = fnCommandAsync
    End Sub

    Public Event CanExecuteChanged As EventHandler Implements ICommand.CanExecuteChanged

    Public Function CanExecute(parameter As Object) As Boolean Implements ICommand.CanExecute
        Return True
    End Function

    Public Sub Execute(parameter As Object) Implements ICommand.Execute
        If Not EAsync Then
            _fnCommand?.Invoke()
        Else
            _fnCommandAsync?.Invoke().Wait()
        End If
    End Sub
End Class

File: ReactlikeMvvm/HiCustomCommand/HiCommand.cs
using System;
using System.Threading.Tasks;
using System.Windows.Input;

namespace ReactlikeMvvm.HiCustomCommand
{
    public class HiCommand : ICommand
    {
        private Action? _fnCommand;
        private Func<Task>? _fnCommandAsync;
        public bool EAsync => _fnCommand is null;
        public HiCommand(Action fnCommand)
        {
            _fnCommand = fnCommand;
        }
        public HiCommand(Func<Task> fnCommandAsync)
        {
            _fnCommandAsync = fnCommandAsync;
        }
        public event EventHandler? CanExecuteChanged;
        public bool CanExecute(object? parameter) => true;

        public void Execute(object? parameter)
        {
            if (!EAsync)
            {
                _fnCommand?.Invoke();
            }
            else
            {
                _fnCommandAsync?.Invoke().Wait();
            }
        }
    }
}